swarmrl/engine/gaurav_sim.py

```diff
@@ -0,0 +1,700 @@
+        force and torque data. Specify path **without** the file extension.
+        The shelve module will find all files needed to load the database.
+    """
+
+    ureg: pint.UnitRegistry
+    box_length: float
+    time_step: float
+    time_slice: float
+    snapshot_interval: float
+    raft_radius: float
+    raft_repulsion_strength: float
+    dynamic_viscosity: float
+    fluid_density: float
+    lubrication_threshold: float
+    magnetic_constant: float
+    capillary_force_data_path: pathlib.Path
+
+
+def setup_unit_system(ureg: pint.UnitRegistry):
+    """
+    In-place definition of the unit system for the simulation,
+    added to the ureg object.
+    """
+    # basis units
+    ureg.define("sim_length = 1 micrometer")
+    ureg.define("sim_time = 1 second")
+    ureg.define("sim_current = 1 ampere")
+    ureg.define("sim_force = 1e-6 newton")
+
+    # gaurav's original unit system isn't consistent, but this should
+    # only affect the water density
+
+    # derived units
+    ureg.define("sim_mass = sim_force * sim_time**2 / sim_length")
+    ureg.define("sim_velocity = sim_length / sim_time")
+    ureg.define("sim_angular_velocity = 1 / sim_time")
+    ureg.define("sim_dyn_viscosity = sim_mass / (sim_length * sim_time)")
+    ureg.define("sim_kin_viscosity = sim_length**2 / sim_time")
+    ureg.define("sim_torque = sim_length * sim_force")
+    ureg.define("sim_magnetic_field = sim_mass / sim_time**2 / sim_current")
+    ureg.define("sim_magnetic_permeability = sim_force / sim_current **2")
+    ureg.define("sim_magnetic_moment = sim_current * sim_length**2")
+
+    return ureg
+
+
+def convert_params_to_simunits(params: GauravSimParams):
+    ureg = params.ureg
+    params_simunits = GauravSimParams(
+        ureg=ureg,
+        box_length=params.box_length.m_as("sim_length"),
+        time_step=params.time_step.m_as("sim_time"),
+        time_slice=params.time_slice.m_as("sim_time"),
+        snapshot_interval=params.snapshot_interval.m_as("sim_time"),
+        raft_radius=params.raft_radius.m_as("sim_length"),
+        raft_repulsion_strength=params.raft_repulsion_strength.m_as("sim_force"),
+        dynamic_viscosity=params.dynamic_viscosity.m_as("sim_dyn_viscosity"),
+        fluid_density=params.fluid_density.m_as("sim_mass / sim_length**3"),
+        lubrication_threshold=params.lubrication_threshold.m_as("sim_length"),
+        magnetic_constant=params.magnetic_constant.m_as("sim_magnetic_permeability"),
+        capillary_force_data_path=params.capillary_force_data_path,
+    )
+    return params_simunits
+
+
+@dataclasses.dataclass
+class Raft:
+    pos: np.array
+    alpha: float  # in radians
+    magnetic_moment: float
+
+    def get_director(self) -> np.array:
+        return np.array([np.cos(self.alpha), np.sin(self.alpha)])
+
+
+@dataclasses.dataclass
+class GauravAction:
+    amplitudes: np.array
+    frequencies: np.array
+    phases: np.array
+    offsets: np.array
+
+
+class Model:
```

Comment:
Is it possible to have these inherit from our `ForceFunction`? It sees like most of the code can be reused we should just rename `colloids` to `particles` or something along those lines. Perhaps we make a more general parent for the force functions so that we can have an ESPResSo and newer ones which then get their own custom type hints and we can do some checks when users try to use espresso actions in an MPI simulation. What do you think?